editor.planx.uk/src/@planx/components/Calculate/Editor.tsx

```diff
@@ -36,12 +33,49 @@ export default function Component(props: Props) {
         props.handleSubmit({ type: TYPES.Calculate, data: newValues });
       }
     },
-    validate: () => {
-      // can parse formula
-      getVariables(formik.values.formula);
+    validate: (values) => {
+      const errors: FormikErrors<Calculate> = {};
+      try {
+        // can parse formula
+        getVariables(formik.values.formula);
+
+        // Validate formula
+        const result = evaluate(
+          values.formula,
+          values.samples,
+          values.defaults,
+        );
+
+        if (Number.isNaN(Number(result))) {
+          errors.formula = "Enter a formula which outputs a number";
+        }
+      } catch (error: any) {
+        errors.formula = error.message;
+      }
+      return errors;
     },
+
+    validateOnChange: false,
   });
 
+  const sampleResult = React.useMemo(() => {
```

Comment:
Tried to ensure this was an additive fix, I didn't want to take away too much which makes the component work because I couldn't fully grasp how it all functioned within the timeboxed time

Reply:
So what I think is happening here is that the `useMemo()` and associated array are just covering the same functionality as `validateOnChange: true`.

The pattern we want to follow, and other components employ, is to keep `validateOnChange: false` and then handle validation onSubmit.

I think this means removing this custom validation function, and just using formik 👍

Reply:
It also feeds the "Try" field which gives you a preview of the formula output:

```tsx
         <p>
            <strong>{formik.values.output || "<output>"}</strong> would be set
            to <strong>{sampleResult}</strong>.
          </p>
```

I could maybe move it to state and use that in the validation step as well, updating with a useEffect or something but, this felt like a change on top of the bug fix. 

Happy to do a bit more on this if you think it is worth it?

<img width="733" alt="Screenshot 2024-10-16 at 11 11 46" src="https://github.com/user-attachments/assets/de45931f-ffc7-4f74-9bed-19004516e194">

Reply:
Aha, thanks. This is what I've missed 👍 

So we do need a way to get a runtime result out which is `sampleResult`.

I think I'd aim to see if there's a way we can avoid duplicating the validation logic - maybe a function with a signature like this could work in both instances?

```ts
type Validate = (values: Calculate) => [result: number | undefined, errors: FormikErrors<Calculate>]
```

This would look something like this - 

```ts
validate: (values) => {
  const [_, errors]: validate(values);
  return errors;
},

...

  const sampleResult = React.useMemo(() => {
    const [result, _]: validate(formik.values);
    if (!result) return UNKNOWN
  }, [formik.values, validate]);
```

Reply:
@DafyddLlyr got to the end of my timebox and didn't get it working as required. It's definitely possible though and I'll make a note for a potential follow-up PR to tidy this up, if that sounds good to you?

I was mostly caught resolving types and harmonising return messaging.